Delete yum artifacts regex

f18m opened the pull request:
To be written

thiagofigueiro commented on src/nexuscli/cli/subcommand_repository.py:
```diff
@@ -24,6 +23,9 @@
          [--write=<w_policy>]
          [--depth=<repo_depth>]
+  nexus3 repository (delete|del) <repo_name> [--force]
+  nexus3 repository del_assets_regex <repo_name> <assets_regex> [--force]
```
Wouldn't these `del_assets*` subcommands be better under the existing root command (`delete|del`)?

```
  nexus3 (delete|del) <repository_path>
```

You could add flags for the user to specify whether the `<repository_path>` contains a regex or a wildcard (glob?)

I'll have a better look at the problem this weekend so I have a better informed suggestion.

f18m replied:
Regarding the 2 points:
1) subcommand vs root command: honestly as user of nexus3-cli I found very confusing having both: indeed it was quite confusing to have these subcommands documented just at the end of the "--help" section and it took me a while to discover how many possibilities nexus3-cli provides beside the root commands.
So IMHO I would completely remove root commands and leave only subcommands: all the list/upload/download/delete root commands IMHO should be under an "assets" subcommand.
Then in this scenario these del_assets_* commands added by this PR could be merged with the "delete" root command perhaps.

2) regarding having a single command instead of del_assets_regex/del_assets_wildcard: I think you're right, having a flag would make it possible to unify them.
Should the default be to interpret the parameter appearing after the <repo_name> as a wildcard or as a regex? Since wildcards are easier to use I would opt for the former.

thiagofigueiro replied:
1. All the commands used to be at the root level but I found that confusing as well. The help page was very long and most users want to upload/download/delete. So I decided to pack them in sub commands as this is already an established pattern in many cli utilities (git, for example).

As changing the CLI would be a breaking change, I’d favour a bit more research/discussion before acting on it. Perhaps build the docopt in a separate branch so we can see how it looks.

2. I agree wildcard makes more sense. There’s an existing behaviour where it deletes everything under a matching name if the name ends with a / (eg del directory/) - if possible, I’d like to keep this behaviour.

Thanks again for your contribution.

f18m replied:
1. Ok sure - I understand that grouping root commands all together would be a breaking change... better not to mix with this MR. So, for now, should I keep the del_assets command under the "repository" subcommand? Or do you prefer to have a new subcommand or a root-level command for that?

2. I have to say that after some more attempts to use the new "del_assets_wildcard" command I discovered that the OrientDB used by Nexus3 is handling wildcards in a weird (at least to me) way: i.e. the wildcard symbol % matches only string prefixes or postfixes. E.g. if I have an asset named "folder1/myasset-1.2.3.rpm" and I provide as wildcard "folder1/%1.2.3%" the Groovy script will report 0 matches. If I provide "%1.2.3%", the asset will be matched.
All this to say: maybe allowing the user to use wildcards instead of regex is not that useful finally...

thiagofigueiro replied:
1. Yes please. I'd prefer to keep it consistent with the subcommand pattern. I'll take your feedback onboard and think of a way to better expose the subcommands to first-time users.

2. I think that's still useful - if we can clearly communicate how the wildcard handling works (or link to a place that does it). One thing that I didn't consider is what happens when a file you want to delete has a wildcard character in it. I guess the user would need to escape it by default, unless we only enable these options through flags; e.g.:

```
  nexus3 (delete|del) <repository_path> [--regex|--wildcard] 

Options:
  --regex          Enable regular expression parsing on <repository_path>
  --wildcard       Enable wildcard parsing on <repository_path>
```

f18m replied:
Hi @thiagofigueiro ,
just to be sure I understand: from reading the example docs you put in point #2 I assume that the answer w.r.t. point #1 is "having these del_assets_wildcard and del_assets_regex" merged with the root-level "del" command right?

f18m replied:
Hi @thiagofigueiro ,
I was going to implement the change but I realized now that regarding point 1
  1. if you want to have the "del_assets_wildcard/regex" commands merged with the "delete|del" command there's a small detail to fix:  former cmds take the repository name and, separated, the repository path. The latter cmd is taking just the repository path. So how are user supposed to delete assets by regex?
Would that be something like:
`
   nexus3 del --regex myrepoName/^[a-z]*myregex.*
`
?
Or can we modify the "delete|del" command to separate the reponame from the asset name:
`
   nexus3 del --regex myrepoName  ^[a-z]*myregex.*
`
?

thiagofigueiro replied:
`--regex` and `--wildcard` would be toggle flags to the `del` command.

There's [existing code](https://github.com/thiagofigueiro/nexus3-cli/blob/master/src/nexuscli/nexus_client.py#L270) that can receive this
```
nexus3 del --regex myrepoName/^[a-z]*myregex.*
```

and split it using something [like this](https://github.com/thiagofigueiro/nexus3-cli/blob/master/src/nexuscli/nexus_client.py#L363).

It's a part of the code that is older and somewhat messy but it does the job.  Assuming this is the docopt line:
```
  nexus3 (delete|del) <repository_path>
```

Then you could do something like this:

```
if options.get('--regex') or options.get('--wildcard'): 
    repository, regex_or_wildcard = self._pop_repository(options.get('<repository_path>'))
    # delete wildcard or regex code
else:
    # existing delete code
```

f18m replied:
Hi @thiagofigueiro ,
I'm trying to do the change but I admit I'm facing several issues as I don't know the code so well.
For example: the "nexus3 (delete|del) <repository_path>" command is not using the "api/repository/*" code, while my PR has added all the code to perform delete operations in the API first. 
Should I change the "delete|del" command to go through the API?
In practice such existing "delete|del" command can be implemented as a special case of the wildcard code where simply no wildcards are provided...

thiagofigueiro commented on src/nexuscli/cli/subcommand_repository.py:
```diff
@@ -129,6 +135,47 @@ def cmd_delete(nexus_client, args):
+            sys.stdout.write('Found 0 matching assets: aborting delete\n')
+            return errors.CliReturnCode.SUCCESS.value
+        
+        sys.stdout.write('Found {} matching assets:\n{}\n'.format(len(assets_list), '\n'.join(assets_list)))
```
Since dropping support for Python 2.x, the whole code-base uses f-strings instead of `.format`. Could you please use f-strings for consistency?

f18m replied:
done!

thiagofigueiro commented:
1. All the commands used to be at the root level but I found that confusing as well. The help page was very long when most users want to upload/download/delete. Do I decided to pack them in sub commands as this is already an established pattern in many cli utilities (git, for example). 

As this would also be a breaking change, I’d favour a bit more research/discussion before acting on it. 

2. I agree wildcard makes more sense. There’s an existing behaviour where it deletes everything under a matching name if the name ends with a / (eg del directory/) - if possible, I’d like to keep this behaviour. 

Thanks again for your contribution. 


Sent on the run; please excuse my brevity.

> On 9 Nov 2019, at 09:44, Francesco Montorsi <notifications@github.com> wrote:
> 
> ﻿
> @f18m commented on this pull request.
> 
> In src/nexuscli/cli/subcommand_repository.py:
> 
> > @@ -24,6 +23,9 @@
>           [--blob=<store_name>] [--strict-content] [--cleanup=<c_policy>]
>           [--write=<w_policy>]
>           [--depth=<repo_depth>]
> +  nexus3 repository (delete|del) <repo_name> [--force]
> +  nexus3 repository del_assets_regex <repo_name> <assets_regex> [--force]
> Regarding the 2 points:
> 
> subcommand vs root command: honestly as user of nexus3-cli I found very confusing having both: indeed it was quite confusing to have these subcommands documented just at the end of the "--help" section and it took me a while to discover how many possibilities nexus3-cli provides beside the root commands.
> So IMHO I would completely remove root commands and leave only subcommands: all the list/upload/download/delete root commands IMHO should be under an "assets" subcommand.
> Then in this scenario these del_assets_* commands added by this PR could be merged with the "delete" root command perhaps.
> 
> regarding having a single command instead of del_assets_regex/del_assets_wildcard: I think you're right, having a flag would make it possible to unify them.
> Should the default be to interpret the parameter appearing after the <repo_name> as a wildcard or as a regex? Since wildcards are easier to use I would opt for the former.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

thiagofigueiro commented:
Hi, Francesco. Sorry I haven’t replied yet - having a busy week at the day job. Please beware with me. I’ll reply tomorrow at latest. 

Sent on the run; please excuse my brevity.

> On 22 Nov 2019, at 08:46, Francesco Montorsi <notifications@github.com> wrote:
> 
> ﻿
> @f18m commented on this pull request.
> 
> In src/nexuscli/cli/subcommand_repository.py:
> 
> > @@ -24,6 +23,9 @@
>           [--blob=<store_name>] [--strict-content] [--cleanup=<c_policy>]
>           [--write=<w_policy>]
>           [--depth=<repo_depth>]
> +  nexus3 repository (delete|del) <repo_name> [--force]
> +  nexus3 repository del_assets_regex <repo_name> <assets_regex> [--force]
> Hi @thiagofigueiro ,
> I was going to implement the change but I realized now that regarding point 1
> 
> if you want to have the "del_assets_wildcard/regex" commands merged with the "delete|del" command there's a small detail to fix: former cmds take the repository name and, separated, the repository path. The latter cmd is taking just the repository path. So how are user supposed to delete assets by regex?
> Would that be something like:
> nexus3 del --regex myrepoName/^[a-z]*myregex.*
> ?
> Or can we modify the "delete|del" command to separate the reponame from the asset name:
> nexus3 del --regex myrepoName ^[a-z]*myregex.*
> ?
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

thiagofigueiro commented:
Yeah that sounds fine. I might write some integration deletion tests to ensure before/after behaves the same. Cheers

Sent on the run; please excuse my brevity.

> On 23 Nov 2019, at 22:51, Francesco Montorsi <notifications@github.com> wrote:
> 
> ﻿
> @f18m commented on this pull request.
> 
> In src/nexuscli/cli/subcommand_repository.py:
> 
> > @@ -24,6 +23,9 @@
>           [--blob=<store_name>] [--strict-content] [--cleanup=<c_policy>]
>           [--write=<w_policy>]
>           [--depth=<repo_depth>]
> +  nexus3 repository (delete|del) <repo_name> [--force]
> +  nexus3 repository del_assets_regex <repo_name> <assets_regex> [--force]
> Hi @thiagofigueiro ,
> I'm trying to do the change but I admit I'm facing several issues as I don't know the code so well.
> For example: the "nexus3 (delete|del) <repository_path>" command is not using the "api/repository/*" code, while my PR has added all the code to perform delete operations in the API first.
> Should I change the "delete|del" command to go through the API?
> In practice such existing "delete|del" command can be implemented as a special case of the wildcard code where simply no wildcards are provided...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

f18m commented:
> Yeah that sounds fine. I might write some integration deletion tests to ensure before/after behaves the same.

ok thanks. I managed to run the tests on my system btw so I might try to write new tests as well and adjust the existing one.
However I have a bigger trouble: when I add --regex or --wildcard to the delete CLI command, the whole cmd is not recognized anymore as a 'root' command and is instead routed to _run_subcommand() function in the CLI main...not sure why though.. I tried looking at docopt but couldn't find any obvious reason for that...

f18m commented:
> However I have a bigger trouble: when I add --regex or --wildcard to the delete CLI command, the whole cmd is not recognized anymore as a 'root' command and is instead routed to _run_subcommand() function in the CLI main...not sure why though.. I tried looking at docopt but couldn't find any obvious reason for that...

Ping in case you have some idea to unblock me :) thanks!

f18m commented:
> > However I have a bigger trouble: when I add --regex or --wildcard to the delete CLI command, the whole cmd is not recognized anymore as a 'root' command and is instead routed to _run_subcommand() function in the CLI main...not sure why though.. I tried looking at docopt but couldn't find any obvious reason for that...
> 
> Ping in case you have some idea to unblock me :) thanks!

btw after some more attempts to get CLI option parsing working I tried the following: 
```
 git clone https://github.com/thiagofigueiro/nexus3-cli.git
 cd nexus3-cli
 pip3 install -e . ; clear ; nexus3 upload foo bar --norecurse
```
and I noticed that also this case does not work: I get the docopt help printed instead of the upload command output.
If I remove the --norecurse option from the invocation it works!

This look like a bug... it is happening on my system with docopt==0.6.2 (latest)...

f18m commented:
Hi @thiagofigueiro ,
I have integrated the CLI fix done recently in master and with that one I was able to test again all this PR. I think now the MR is in a good shape and ready for re-review.

It's still missing integration tests though. I may look into adding them later... for now I conducted a manual testing of all feature (--force, --regex, --wildcard) against a number of repository types/format combinations.

This PR finally brings:
 - 3 different asset matching policies: exact name, regex or wildcard
 - confirmation before delete

In this MR I'm erasing the current "del|delete" command implementation, which does not go through the "api" module and is missing all above features... let me know if that's ok for you.

f18m commented:
> In this MR I'm erasing the current "del|delete" command implementation, which does not go through the "api" module and is missing all above features... let me know if that's ok for you.

ping :)

f18m commented:
> > In this MR I'm erasing the current "del|delete" command implementation
> 
> This should be ok, as long as it all still fits together in a sensible manner. I don't think it's the case here but, if there's an API breaking change, please bump the major version of the project.
> 
> I'll wait until there are passing tests before reviewing this again.

Ok thanks. I have some issues running the integration tests though - I will open a different ticket to avoid polluting this one.

f18m commented:
> > In this MR I'm erasing the current "del|delete" command implementation
> 
> This should be ok, as long as it all still fits together in a sensible manner. I don't think it's the case here but, if there's an API breaking change, please bump the major version of the project.
> 

ok I don't think this will be the case for this MR though

> I'll wait until there are passing tests before reviewing this again.

ok sure I'm looking at the tests these days but I have a couple of doubts about the tests... it would be great to have some little guidance: 

1) is the "tests/cli/test_delete" (which is failing right now) an integration test or not?
It seems to run both when I issue 
`pytest -m 'integration'`
and also when I run 
`pytest -m 'not integration'`

2) in that test is the Nexus response mocked ? Should that test run against the docker container of the Nexus or instead against a mocked Nexus?

3) since with this MR the delete command is changed to go through the API path and use a Groovy script to do the delete, do you think the test should be changed? and if so, can you point me to an example to lay out the new test?

Thanks and sorry for all the questions but I have never used pytest-mock and for me it's more difficult to follow the test code rather than the "production" code :)

Thanks

f18m commented on src/nexuscli/api/repository/collection.py:
```diff
@@ -146,6 +147,36 @@ def delete(self, name):
+        self._client.scripts.create_if_missing(SCRIPT_NAME_DELETE_ASSETS, content)
+        
+        # prepare JSON for Groovy:
+        jsonData = {}
```
fixed

f18m commented on src/nexuscli/api/repository/collection.py:
```diff
@@ -225,6 +227,49 @@ def delete(self, name):
+        # parse the JSON we got back
+        if 'result' not in groovy_returned_json:
+            raise exception.NexusClientAPIError(groovy_returned_json)
+        script_result = json.loads(groovy_returned_json['result']) # this is actually a JSON: convert to Python dict
```
I think that ScriptCollection.run() is already doing .json() on the HTTP response from the Nexus. However I think that the response.result field is interpreted as a string and does not handle the case whether that string is another JSON... hence the json.loads()...

f18m commented on src/nexuscli/cli/root_commands.py:
```diff
@@ -137,16 +140,62 @@ def cmd_dl(*args, **kwargs):
+    nl = '\n' # see https://stackoverflow.com/questions/44780357/how-to-use-newline-n-in-f-string-to-format-output-in-python-3-6
+    
+    if not doForce:
+        sys.stdout.write(f'Retrieving assets matching {assetMatchOption.name} "{assetName}" from repository "{repoName}"\n')
```
ok fixed

f18m commented on src/nexuscli/api/repository/collection.py:
```diff
@@ -225,6 +227,49 @@ def delete(self, name):
+        if script_result == None or 'assets' not in script_result:
+            raise exception.NexusClientAPIError(groovy_returned_json)
+        
+        if 'success' in script_result and script_result['success']==False:
```
ok fixed

f18m commented on src/nexuscli/cli/root_commands.py:
```diff
@@ -137,16 +140,62 @@ def cmd_dl(*args, **kwargs):
+            return errors.CliReturnCode.API_ERROR.value
+            
+        if len(assets_list) == 0:
+            sys.stdout.write('Found 0 matching assets: aborting delete\n')
```
ok I changed that - however I'm not sure why "stderr" is used here... if that's not an error it should probably go in stdout like other messages...

f18m commented on src/nexuscli/cli/root_commands.py:
```diff
@@ -137,16 +140,62 @@ def cmd_dl(*args, **kwargs):
+            return errors.CliReturnCode.SUCCESS.value
+        
+        sys.stdout.write(f'Found {len(assets_list)} matching assets:\n{nl.join(assets_list)}\n')
+        util.input_with_default(
```
well the --force option is still considered: a few lines above there's an `if not doForce:`. That if is used to skip not only the interactive question but the whole process of first finding the matching list doing a dry run and then, if the confirmation is provided by the user, to actually delete.
If --force is given, we immediately ask the Groovy script to proceed with the delete...

f18m commented on src/nexuscli/cli/root_commands.py:
```diff
@@ -137,16 +140,62 @@ def cmd_dl(*args, **kwargs):
+    assetMatch = AssetMatchOptions.EXACT_NAME
+    if options.get('--wildcard') and options.get('--regex'):
+        sys.stderr.write(
+            f'Cannot provide both --regex and --wildcard\n')
```
fixed

f18m commented on src/nexuscli/cli/root_commands.py:
```diff
@@ -137,16 +140,62 @@ def cmd_dl(*args, **kwargs):
+def _cmd_del_assets(nexus_client, repoName, assetName, assetMatchOption, doForce):
+    """Performs ``nexus3 repository delete_assets``"""
+    
+    nl = '\n' # see https://stackoverflow.com/questions/44780357/how-to-use-newline-n-in-f-string-to-format-output-in-python-3-6
```
right - was used in a previous attempt - deleted now

f18m replied:
actually this is used: in 2 print() commands to show the list of deleted files I'm doing: `nl.join(assets_list)` to get a newline-separated list of files...

bt-thiago commented:
>     1. is the "tests/cli/test_delete" (which is failing right now) an integration test or not?

Integration tests have a `@pytest.mark.integration` decorator.

> in that test is the Nexus response mocked ? Should that test run against the docker container of the Nexus or instead against a mocked Nexus?

It's a unit test (not integration) so the response is mocked.  There isn't a integration test for this because I don't need to test Nexus itself - I call their API and trust that they do the right thing.

However, since you're introducing a Groovy script to perform the deletion, we can no longer trust that the server side is doing the "right thing", so it would be a good idea to introduce integration tests.

> ... do you think the test should be changed? and if so, can you point me to an example to lay out the new test?

As above, yes. We should keep a unit test to validate the internal API interface and add integration tests to validate the "nexus API" (ie the groovy script) interface and implementation.

> Thanks and sorry for all the questions but I have never used pytest-mock and for me it's more difficult to follow the test code rather than the "production" code :)

No worries; mocking can get complicated. You have spent considerable time on this - I appreciate it and I hope it's being useful as a learning experience. Ultimately I'm responsible for maintaining this code so please understand if I don't merge this straight-away once the tests are passing as I will probably want to review the code mode in depth and do some refactoring along the way.

f18m commented:
> It's a unit test (not integration) so the response is mocked. There isn't a integration test for this because I don't need to test Nexus itself - I call their API and trust that they do the right thing.

actually I think I found out that there is an integration test with exactly the same name (that's what confused me at the beginning) inside the tests/cli/test_cli.py file.
I think that it was an integration test removing the whole contents of a folder tree - this is a feature that the groovy script of this MR currently does not support, so I modified the test to delete file by file (and added verification that at the end the repo is empty).

> However, since you're introducing a Groovy script to perform the deletion, we can no longer trust that the server side is doing the "right thing", so it would be a good idea to introduce integration tests.
> 
Right - I spent more time on the integration test rather than the unit test because I think in this case it's mostly useful the integration test.

> No worries; mocking can get complicated. You have spent considerable time on this - I appreciate it and I hope it's being useful as a learning experience. Ultimately I'm responsible for maintaining this code so please understand if I don't merge this straight-away once the tests are passing as I will probably want to review the code mode in depth and do some refactoring along the way.

Sure right - just let me know if I can do something else to get this integrated. The thing is: in my company I'm introducing Nexus as new technology to store artifacts and we really need a tool like nexus3-cli to operate on the artefacts we store there.
As we are in a sort of hurry I may try to build a Pypi package from my fork of nexus3-cli till this MR is fully integrated (also because we will need to add support for Docker repos to nexus3 shortly).
This way I will be able to decouple this MR from the timelines of my needs...

Thanks!
